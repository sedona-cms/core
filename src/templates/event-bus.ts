import Vue from 'vue'

const vueInstance = new Vue()
const eventBus = {
  emit(event: string, ...args: any[]): void {
    vueInstance.$emit(event, args)
  },
  on(event: string | string[], callback: Function): void {
    vueInstance.$on(event, callback)
  },
  off(event?: string | string[], callback?: Function): void {
    vueInstance.$off(event, callback)
  },
  once(event: string | string[], callback: Function): void {
    vueInstance.$once(event, callback)
  },
}

export { eventBus }
