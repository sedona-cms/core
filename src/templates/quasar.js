import VuePlugin from 'quasar/src/vue-plugin.js'

import {QToolbar, QToolbarTitle} from 'quasar/src/components/toolbar'
import {QSpace} from 'quasar/src/components/space'
import {QBtn} from 'quasar/src/components/btn'
import {QBtnDropdown} from 'quasar/src/components/btn-dropdown'
import {QIcon} from 'quasar/src/components/icon'
import {QTabPanel, QTabPanels} from 'quasar/src/components/tab-panels'
import {QList, QItem, QItemSection, QItemLabel} from 'quasar/src/components/item'
import {QExpansionItem} from 'quasar/src/components/expansion-item'
import {QAvatar} from 'quasar/src/components/avatar'
import {QSeparator} from 'quasar/src/components/separator'
import {QTooltip} from 'quasar/src/components/tooltip'
import {QDialog} from 'quasar/src/components/dialog'
import {QCard} from 'quasar/src/components/card'
import {QTable, QTd, QTh, QTr} from 'quasar/src/components/table'
import {QForm} from 'quasar/src/components/form'
import {QInput} from 'quasar/src/components/input'
import {QCircularProgress} from 'quasar/src/components/circular-progress'
import {QMenu} from 'quasar/src/components/menu'
import {QOptionGroup} from 'quasar/src/components/option-group'
import {QField} from 'quasar/src/components/field'

import { QScrollArea } from 'quasar/src/components/scroll-area'

import * as directives from 'quasar/src/directives.js'
import {Dialog} from 'quasar/src/plugins.js'

export default {
    ...VuePlugin,
    install(Vue) {
        VuePlugin.install(Vue, {
            components: {
                QToolbar,
                QToolbarTitle,
                QSpace,
                QBtn,
                QBtnDropdown,
                QIcon,
                QTabPanel,
                QTabPanels,
                QList,
                QItem,
                QItemSection,
                QItemLabel,
                QExpansionItem,
                QAvatar,
                QSeparator,
                QTooltip,
                QDialog,
                QCard,
                QTable,
                QTd,
                QTh,
                QTr,
                QForm,
                QInput,
                QCircularProgress,
                QMenu,
                QOptionGroup,
                QField,
                QScrollArea,
            },
            directives,
            plugins: {
                Dialog,
            },
        })
    },
}
