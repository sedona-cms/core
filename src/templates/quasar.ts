import VuePlugin from 'quasar/src/vue-plugin'

import { QToolbar, QToolbarTitle } from 'quasar/src/components/toolbar'
import { QSpace } from 'quasar/src/components/space'
import { QBtn } from 'quasar/src/components/btn'
import { QBtnDropdown } from 'quasar/src/components/btn-dropdown'
import { QIcon } from 'quasar/src/components/icon'
import { QTabPanel, QTabPanels } from 'quasar/src/components/tab-panels'
import { QList, QItem, QItemSection, QItemLabel } from 'quasar/src/components/item'
import { QExpansionItem } from 'quasar/src/components/expansion-item'
import { QAvatar } from 'quasar/src/components/avatar'
import { QSeparator } from 'quasar/src/components/separator'
import { QTooltip } from 'quasar/src/components/tooltip'

import { QCard, QCardSection, QCardActions } from 'quasar/src/components/card'

import { QImg } from 'quasar/src/components/img'
import { QVideo } from 'quasar/src/components/video'

import { QInput } from 'quasar/src/components/input'
import { QCheckbox } from 'quasar/src/components/checkbox'
import { QField } from 'quasar/src/components/field'
import { QDate } from 'quasar/src/components/date'
import { QEditor } from 'quasar/src/components/editor'

import { QCircularProgress } from 'quasar/src/components/circular-progress'
import { QMenu } from 'quasar/src/components/menu'
import { QOptionGroup } from 'quasar/src/components/option-group'

import { QScrollArea } from 'quasar/src/components/scroll-area'

import * as directives from 'quasar/src/directives'

export default {
  ...VuePlugin,
  install(Vue) {
    VuePlugin.install(Vue, {
      components: {
        QToolbar,
        QToolbarTitle,
        QSpace,
        QBtn,
        QBtnDropdown,
        QIcon,
        QTabPanel,
        QTabPanels,
        QList,
        QItem,
        QItemSection,
        QItemLabel,
        QExpansionItem,
        QAvatar,
        QSeparator,
        QTooltip,
        QCard,
        QCardSection,
        QCardActions,
        QInput,
        QCheckbox,
        QDate,
        QEditor,
        QCircularProgress,
        QMenu,
        QOptionGroup,
        QField,
        QScrollArea,
        QImg,
        QVideo,
      },
      directives,
      plugins: {},
    })
  },
}
