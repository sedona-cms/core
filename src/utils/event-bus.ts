import Vue from 'vue'

type EventCallback = (...args: never[]) => void

const vueInstance = new Vue()
const eventBus = {
  emit(event: string, ...args: unknown[]): void {
    vueInstance.$emit(event, args)
  },
  on(event: string | string[], callback: EventCallback): void {
    vueInstance.$on(event, callback)
  },
  off(event?: string | string[], callback?: EventCallback): void {
    vueInstance.$off(event, callback)
  },
  once(event: string | string[], callback: EventCallback): void {
    vueInstance.$once(event, callback)
  },
}

export { eventBus }
