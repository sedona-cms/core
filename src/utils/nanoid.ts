import { customAlphabet } from 'nanoid'

const alphabet = '0123456789abcdefghijklmnopqrstuvwxyz'
const nanoid = customAlphabet(alphabet, 10)

/**
 * Generate unique identifier
 *
 * @exports
 * @returns {string} short unique id
 */
export function generateId(): string {
  return nanoid()
}
