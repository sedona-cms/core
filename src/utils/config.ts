import path from 'path'
import fs from 'fs'
import Ajv from 'ajv'
import { generateId } from './nanoid'

import menuItemConditionSchema from '../schema/menu-item-condition.schema.json'
import menuItemSchema from '../schema/menu-item.schema.json'
import configSchema from '../schema/sedona.config.schema.json'
import savePanelSchema from '../schema/save-panel.schema.json'

/**
 * Load sedona.config.json from root directory
 *
 * @param {string} rootPath path to src directory
 * @returns {Promise<ModuleConfig>} config object
 */
export async function loadConfigFile(rootPath: string): Promise<ModuleConfig> {
  const configFile = path.resolve(rootPath, 'sedona.config.json')
  if (fs.existsSync(configFile)) {
    const data = await require(`${configFile}`)
    validateConfigFile(data)
    return JSON.parse(JSON.stringify(data))
  }
  throw new Error(`Sedona CMS config file not found in ${configFile}`)
}

export function setIdToMenuItems(items: MenuItem[]): MenuItem[] {
  return items.map(item => {
    if (item.id === undefined) {
      item.id = generateId()
    }
    if (item.type === 'section' && Array.isArray(item.items)) {
      item.items = setIdToMenuItems(item.items)
    }
    return item
  })
}

function validateConfigFile(data: string): void {
  const ajv = new Ajv({
    extendRefs: true,
  })
  ajv.addMetaSchema(menuItemSchema, 'menu-item.schema.json')
  ajv.addMetaSchema(menuItemConditionSchema, 'menu-item-condition.schema.json')
  ajv.addMetaSchema(savePanelSchema, 'save-panel.schema.json')

  const validate = ajv.compile(configSchema)
  const valid = validate(data)

  if (!valid && Array.isArray(validate.errors)) {
    const message = ['Schema Config Validation Error']
    validate.errors.forEach(error => {
      message.push(`data: ${error.dataPath}`)
      message.push(`schema: ${error.schemaPath}`)
      message.push(`message: ${error.message}`)
    })

    throw new Error(message.join(`\n`))
  }
}
