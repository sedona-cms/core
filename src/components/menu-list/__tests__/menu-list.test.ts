import { mount, createLocalVue } from '@vue/test-utils'
import { Quasar, QItemLabel, QItemSection, QIcon, QItem, QAvatar, QList } from 'quasar'
import { MenuList } from '..'
import { generateId } from '../../../utils/nanoid'
import { eventBus } from '../../../utils/event-bus'

const localVue = createLocalVue()

localVue.use(Quasar, {
  components: {
    QItemLabel,
    QItemSection,
    QIcon,
    QItem,
    QAvatar,
    QList,
  },
})

describe('MenuList', () => {
  test('it can be exists with no menu items', () => {
    const wrapper = mount(MenuList, {
      localVue,
      propsData: {
        items: [],
      },
    })
    const vm = wrapper.vm

    expect(wrapper.exists()).toBeTruthy

    expect(vm.menuItems).toBeInstanceOf(Array)
    expect(vm.menuItems).toHaveLength(0)
  })

  test('it can be render menu items', async () => {
    const items = [
      {
        id: generateId(),
        title: 'Post List',
        type: 'item',
        component: 'posts/posts-view',
      },
      {
        id: generateId(),
        title: 'SEO',
        type: 'section',
        items: [],
      },
    ]

    const wrapper = mount(MenuList, {
      localVue,
      propsData: {
        items,
      },
    })
    const vm = wrapper.vm

    expect(wrapper.exists()).toBeTruthy
    expect(vm.menuItems).toBeInstanceOf(Array)
    expect(vm.menuItems).toHaveLength(2)

    expect(wrapper.html()).toMatchSnapshot()

    const labelWrapper = wrapper.findAll('.q-item__label')
    expect(labelWrapper).toHaveLength(2)
    expect(labelWrapper.at(0).text()).toEqual('Post List')
    expect(labelWrapper.at(1).text()).toEqual('SEO')

    expect(wrapper.get('.q-avatar__content').text()).toEqual('folder')

    eventBus.on('core:navigate', ([item]) => {
      expect(item).toBeInstanceOf(Object)
      expect(item).toHaveProperty('id', items[0].id)
      expect(item).toHaveProperty('title', items[0].title)
      expect(item).toHaveProperty('component', items[0].component)
      expect(item).toHaveProperty('subTitle', '')
      expect(item).toHaveProperty('icon', 'folder')
      expect(item).toHaveProperty('type', 'item')
      expect(item).toHaveProperty('params', {})
    })

    wrapper.find('.q-item').trigger('click')
  })

  test('it can be render menu header', () => {
    const items = [
      {
        id: generateId(),
        title: 'Post Meta',
        type: 'header',
      },
    ]
    const wrapper = mount(MenuList, {
      localVue,
      propsData: {
        items,
      },
    })
    const vm = wrapper.vm

    expect(wrapper.exists()).toBeTruthy

    expect(vm.menuItems).toBeInstanceOf(Array)
    expect(vm.menuItems).toHaveLength(1)

    expect(wrapper.get('.q-item__label').text()).toEqual('Post Meta')
  })

  test('it can be render menu section', () => {
    const items = [
      {
        id: generateId(),
        title: 'Post Meta',
        subTitle: 'Post Meta Subtitle',
        type: 'section',
        items: [
          {
            id: generateId(),
            title: 'Post List',
            type: 'item',
            component: 'posts/posts-view',
          },
          {
            id: generateId(),
            title: 'SEO',
            type: 'section',
            items: [],
          },
        ],
      },
    ]
    const wrapper = mount(MenuList, {
      localVue,
      propsData: {
        items,
      },
    })
    const vm = wrapper.vm

    expect(vm.menuItems).toBeInstanceOf(Array)
    expect(vm.menuItems).toHaveLength(1)
  })
})
