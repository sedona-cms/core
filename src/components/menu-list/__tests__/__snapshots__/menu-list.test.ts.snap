// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuList it can be render menu items 1`] = `
"<div class=\\"q-list q-list--dark q-list--padding full-width\\">
  <div tabindex=\\"0\\" class=\\"q-item q-item-type row no-wrap q-item--clickable q-link cursor-pointer q-focusable q-hoverable\\">
    <div tabindex=\\"-1\\" class=\\"q-focus-helper\\"></div>
    <div class=\\"q-item__section column q-item__section--avatar q-item__section--side justify-center\\">
      <div class=\\"q-avatar\\">
        <div class=\\"q-avatar__content row flex-center overflow-hidden bg-grey-7 text-white q-chip--colored\\"><i aria-hidden=\\"true\\" role=\\"presentation\\" class=\\"material-icons q-icon notranslate\\">folder</i></div>
      </div>
    </div>
    <div class=\\"q-item__section column q-item__section--main justify-center\\">
      <div class=\\"q-item__label ellipsis\\">Post List</div>
    </div>
  </div>
  <div tabindex=\\"0\\" class=\\"q-item q-item-type row no-wrap q-item--clickable q-link cursor-pointer q-focusable q-hoverable\\">
    <div tabindex=\\"-1\\" class=\\"q-focus-helper\\"></div>
    <div class=\\"q-item__section column q-item__section--avatar q-item__section--side justify-center\\">
      <div class=\\"q-avatar\\">
        <div class=\\"q-avatar__content row flex-center overflow-hidden bg-grey-7 text-white q-chip--colored\\"><i aria-hidden=\\"true\\" role=\\"presentation\\" class=\\"material-icons q-icon notranslate\\">folder</i></div>
      </div>
    </div>
    <div class=\\"q-item__section column q-item__section--main justify-center\\">
      <div class=\\"q-item__label ellipsis\\">SEO</div>
    </div>
    <div class=\\"q-item__section column q-item__section--side justify-center\\"><i aria-hidden=\\"true\\" role=\\"presentation\\" class=\\"material-icons q-icon notranslate\\">keyboard_arrow_right</i></div>
  </div>
</div>"
`;
